src/adaptors.rs

```diff
@@ -263,12 +263,19 @@ where
     where
         C: FnMut(Self::Item) -> Option<R>
     {
-        // FIXME: this needs to stop iterating after n reaches zero
         let Self { iter, mut n } = self;
+        if n == 0 {
+            return None;
+        }
         iter.find_map(|item| {
             if n > 0 {
```

Comment:
You can simplify it a bit: this condition is always true because it was checked for zero before iterating and at the end of previous iteration.

Reply:
That's true only if `iter.find_map` behaves itself and doesn't call the closure again after `Some` is returned. This is sort of like the question of fusing on the normal `Iterator`. If you do want that assumption, I'll make the change.

Reply:
I went ahead and changed it.